### ISSUE TITLE
Performance Regression Analysis: Data Protocol Changes

### ISSUE THREAD

**Issue by pingu-minimax** [pingu-minimax/EasyR1-Investigation#1]
# Performance Regression Analysis: Data Protocol Changes

## Overview
This issue tracks the investigation into potential performance regressions introduced by recent data protocol changes in commit `098931530606d22f867fd121b1dcb3225a43661f`.

## Background
Based on user reports in issues #39 and #41, we have identified potential performance degradation that may be linked to changes in the data protocol implementation.

## Investigation Tracks

We have created three investigation branches to systematically analyze different aspects:

1. **`investigate-protocol-changes`** - Testing protocol-related performance issues
2. **`investigate-batch-processing`** - Testing batch processing performance issues
3. **`investigate-memory-usage`** - Testing memory utilization performance issues

## Key Files Under Investigation

- `verl/protocol.py` - Core data protocol implementation
- `examples/config.yaml` - Configuration settings

## Sub-Issues

This tracking issue will have the following sub-issues attached for detailed investigation:

- [ ] Test Performance Impact: fix multi modal data oom
- [ ] Test Performance Impact: upgrade vllm to 0.10
- [ ] Test Performance Impact: non blocking false by default

## Timeline
- Investigation started: 2026-02-01
- Target completion: TBD

## Related Issues
- #39
- #41

**Comment by pingu-minimax** [comment]
## File Change Analysis: `verl/protocol.py`

### Commit Reference
- **Commit SHA:** `098931530606d22f867fd121b1dcb3225a43661f`
- **File:** `verl/protocol.py`

### Key Changes Identified

The following changes in `verl/protocol.py` may be causing performance regressions:

#### 1. `DataProto.to()` Method - Non-Blocking Default Change
```python
def to(self, device: torch.device, non_blocking: bool = False) -> "DataProto":
    """Move the batch to device

    Args:
        device (torch.device): the device to move to.
        non_blocking (bool, optional): whether to use non-blocking mode. Defaults to False.

    Returns:
        DataProto: the current DataProto.

    NOTE: remember to use torch.cuda.synchronize() after self.to("cpu") to avoid weird number
    """
    if self.batch is not None:
        self.batch = self.batch.to(device, non_blocking=non_blocking)

    return self
```

**Impact Analysis:**
- The `non_blocking=False` default may cause synchronous GPU-CPU transfers
- This can block execution and reduce throughput during data movement
- Previous behavior may have used `non_blocking=True` for async transfers

#### 2. Serialization and Memory Management
The commit may have modified how `TensorDict` objects are serialized and consolidated:
- `__getstate__` method uses `batch_to_save.consolidate()`
- `__setstate__` loads data with `weights_only=False`

### Next Steps
- Profile GPU utilization during data transfers
- Measure impact of `non_blocking` parameter on training throughput
- Test with explicit `non_blocking=True` to compare performance

**Comment by pingu-minimax** [comment]
## File Change Analysis: `examples/config.yaml`

### Commit Reference
- **Commit SHA:** `098931530606d22f867fd121b1dcb3225a43661f`
- **File:** `examples/config.yaml`

### Key Configuration Changes Identified

The following configuration changes in `examples/config.yaml` may be contributing to performance regressions:

#### 1. Micro Batch Size Changes
```yaml
worker:
  actor:
    micro_batch_size_per_device_for_update: 1   # CHANGED (was 4)
    micro_batch_size_per_device_for_experience: 2  # CHANGED (was 16)
```

**Impact Analysis:**
- **`micro_batch_size_per_device_for_update: 1`** (previous: 4)
  - Reduces batch efficiency during model updates
  - May increase gradient accumulation overhead
  - More frequent smaller updates can slow down training

- **`micro_batch_size_per_device_for_experience: 2`** (previous: 16)
  - Dramatically reduced from 16 to 2 (87.5% reduction)
  - Experience collection now processes much smaller batches
  - May cause more frequent GPU kernel launches
  - Increased CPU-GPU synchronization overhead

#### 2. Memory-Related Configuration
```yaml
worker:
  rollout:
    gpu_memory_utilization: 0.6
```

This setting limits GPU memory to 60%, which combined with smaller batch sizes could indicate an OOM fix that sacrificed throughput.

### Correlation with User Reports
- Issue #39 and #41 report performance degradation
- The micro batch size reductions are likely the primary cause
- These changes appear to be a conservative fix for OOM issues

### Recommendations
1. Test intermediate values: `micro_batch_size_per_device_for_update: 2` and `micro_batch_size_per_device_for_experience: 8`
2. Profile memory usage to find optimal batch sizes
3. Consider dynamic batch sizing based on available GPU memory
4. Investigate if gradient checkpointing can allow larger batches
